README.md
# 🍵 Matcha

<!-- MODULEDOC BLURB -->

> **_First-class match specifications for Elixir._**

<!-- MODULEDOC BLURB -->

[![Version][hex-pm-version-badge]][hex-pm-versions]
[![Documentation][docs-badge]][docs]
[![Dependencies][deps-badge]][deps]
[![License][hex-pm-license-badge]][hex-pm-package]
[![Benchmarks][benchmarks-badge]][benchmarks]
[![Contributors][contributors-badge]][contributors]

|         👍         |                  [Test Suite][suite]                  |                   [Test Coverage][coverage]                    |
| :----------------: | :---------------------------------------------------: | :------------------------------------------------------------: |
| [Release][release] | [![Build Status][release-suite-badge]][release-suite] | [![Coverage Status][release-coverage-badge]][release-coverage] |
|  [Latest][latest]  |  [![Build Status][latest-suite-badge]][latest-suite]  |  [![Coverage Status][latest-coverage-badge]][latest-coverage]  |

## Usage

### Installation

`Matcha` is distributed via [hex.pm][hex-pm], you can install it with your dependency manager of choice using the config provided on its [hex.pm package][hex-pm-package] listing.

<!-- MODULEDOC SNIPPET -->
<!--
  all hyperlinks in this snippet must be inline,
  rather than using markdown link references
-->

### Synopsis

`Matcha` offers tight integration with Elixir and match specifications.

Match specifications are a BEAM VM feature that can execute de-structuring, pattern matching, and re-structring operations very close-to-the-metal. They can be used to efficiently:

- [filter/map in-memory data](https://www.erlang.org/doc/man/ets.html#match_spec_run-2)
- [find ETS objects](https://erlang.org/doc/man/ets.html#select-2)
- [trace specific function calls](https://erlang.org/doc/man/dbg.html#tp-2)

However, they are notoriously difficult to compose and use. Matcha makes this intuitive with ergonomic macros to compose them, and a high-level API with which to use them.

### Examples

```elixir
# Turn Elixir code into a match specification,
#  then use it to filter/map some data
iex> require Matcha
...> Matcha.spec do
...>   {x, y, z} -> x + y + z
...> end
...> |> Matcha.Spec.run!([
...>   {1, 2, 3},
...>   {1, 2},
...>   {1, 2, 3, 4},
...>   {4, 5, 6}
...> ])
[6, 15]
```

This is one way to run test and develop match specifications, but they truly shine in table and tracing applications!

### Guides

Check out [the interactive usage guides](https://hexdocs.pm/matcha/guide-usage.html#content), including using Matcha for:

- [filtering & mapping data](https://hexdocs.pm/matcha/guide-filtering-and-mapping.html#content)
- [selecting objects from tables](https://hexdocs.pm/matcha/guide-tables.html#content)
- [tracing function calls](https://hexdocs.pm/matcha/guide-tracing.html#content)

<!-- MODULEDOC SNIPPET -->

### Documentation

Complete documentation, including guides, are hosted online on [hexdocs.pm][docs].

## Contributing

Contributions are welcome! Check out the [contributing guide][contributing] for more information, and suggestions on where to start.

## Supported Versions

Matcha strives to support the most recent three versions of Elixir and the Erlang/OTPs they support. The canonical list of supported combinations is available by checking the latest successful [test matrix run][test-matrix], but boils down to:

- Elixir 1.12.x
  - OTP 22.x
  - OTP 23.x
  - OTP 24.x
- Elixir 1.13.x
  - OTP 22.x
  - OTP 23.x
  - OTP 24.x
  - OTP 25.x
- Elixir 1.14.x
  - OTP 23.x
  - OTP 24.x
  - OTP 25.x

Earlier versions of Elixir tend to work well, back to Elixir 1.10, but the test suite often uses newer syntax that is complicated to circumvent when testing on older versions (ex. range step literals), so we don't officially commit to being compatible with them since they are not a part of the suite. Versions that may work include:

- Elixir 1.10.x
  - OTP 21.x
  - OTP 22.x
  - OTP 23.x
- Elixir 1.11.x
  - OTP 21.x
  - OTP 22.x
  - OTP 23.x
  - OTP 24.x

Since `Matcha` pokes around in compiler internals, it is important to get ahead of upcoming changes to the language.
This is accomplished with [nightly builds][test-edge] against the latest versions of Elixir, Erlang/OTP, and dependencies;
which catches issues like [internal compiler function signature changes](https://github.com/christhekeele/matcha/commit/27f3f34284349d807fcd2817a04cb4628498a7eb#diff-daf93cf4dc6034e9862d0d844c783586210ea822ae6ded51d925b0ac9e09766bR31-R43)
well in advance of release.

<!-- LINKS & IMAGES -->

<!-- Hex -->

[hex-pm]: https://hex.pm
[hex-pm-package]: https://hex.pm/packages/matcha
[hex-pm-versions]: https://hex.pm/packages/matcha/versions
[hex-pm-version-badge]: https://img.shields.io/hexpm/v/matcha.svg?cacheSeconds=86400&style=flat-square
[hex-pm-downloads-badge]: https://img.shields.io/hexpm/dt/matcha.svg?cacheSeconds=86400&style=flat-square
[hex-pm-license-badge]: https://img.shields.io/badge/license-MIT-7D26CD.svg?cacheSeconds=86400&style=flat-square

<!-- Docs -->

[docs]: https://hexdocs.pm/matcha/index.html
[docs-guides]: https://hexdocs.pm/matcha/usage.html#content
[docs-badge]: https://img.shields.io/badge/documentation-online-purple?cacheSeconds=86400&style=flat-square

<!-- Deps -->

[deps]: https://hex.pm/packages/matcha
[deps-badge]: https://img.shields.io/badge/dependencies-1-blue?cacheSeconds=86400&style=flat-square

<!-- Benchmarks -->

[benchmarks]: https://christhekeele.github.io/matcha/bench
[benchmarks-badge]: https://img.shields.io/badge/benchmarks-online-2ab8b5?cacheSeconds=86400&style=flat-square

<!-- Contributors -->

[contributors]: https://hexdocs.pm/matcha/contributors.html
[contributors-badge]: https://img.shields.io/badge/contributors-%F0%9F%92%9C-lightgrey

<!-- Status -->

[suite]: https://github.com/christhekeele/matcha/actions?query=workflow%3A%22Test+Suite%22
[coverage]: https://coveralls.io/github/christhekeele/matcha

<!-- Release Status -->

[release]: https://github.com/christhekeele/matcha/tree/release
[release-suite]: https://github.com/christhekeele/matcha/actions?query=workflow%3A%22Test+Suites%22+branch%3Arelease
[release-suite-badge]: https://img.shields.io/github/actions/workflow/status/christhekeele/matcha/test-suite.yml?branch=release&cacheSeconds=86400&style=flat-square
[release-coverage]: https://coveralls.io/github/christhekeele/matcha?branch=release
[release-coverage-badge]: https://img.shields.io/coverallsCoverage/github/christhekeele/matcha?branch=release&cacheSeconds=86400&style=flat-square

<!-- Latest Status -->

[latest]: https://github.com/christhekeele/matcha/tree/latest
[latest-suite]: https://github.com/christhekeele/matcha/actions?query=workflow%3A%22Test+Suites%22+branch%3Alatest
[latest-suite-badge]: https://img.shields.io/github/actions/workflow/status/christhekeele/matcha/test-suite.yml?branch=latest&cacheSeconds=86400&style=flat-square
[latest-coverage]: https://coveralls.io/github/christhekeele/matcha?branch=latest
[latest-coverage-badge]: https://img.shields.io/coverallsCoverage/github/christhekeele/matcha?branch=latest&cacheSeconds=86400&style=flat-square

<!-- Other -->

[elixir-version-requirements]: https://hexdocs.pm/elixir/Version.html#module-requirements
[changelog]: https://hexdocs.pm/matcha/changelog.html
[test-matrix]: https://github.com/christhekeele/matcha/actions/workflows/test-matrix.yml
[test-edge]: https://github.com/christhekeele/matcha/actions/workflows/test-edge.yml
[contributing]: https://hexdocs.pm/matcha/contributing.html


CHANGELOG.md
# Changelog

Matcha uses [Semantic Versioning 2.0.0](https://semver.org/).

Changes before `v0.2.0` are considered pre-release and not included here.


mix.exs
defmodule Matcha.MixProject do
  use Mix.Project

  @name "Matcha"
  @description "First-class match specification and match patterns for Elixir"
  @authors ["Chris Keele"]
  @maintainers ["Chris Keele"]
  @licenses ["MIT"]

  @release_branch "release"

  @github_url "https://github.com/christhekeele/matcha"
  @homepage_url @github_url

  @dev_envs [:dev, :test]
  @default_test_suite_includes "--include doctest --include unit --include usage"

  def project,
    do: [
      # Application
      app: :matcha,
      elixir: "~> 1.10",
      start_permanent: Mix.env() == :prod,
      version: "VERSION" |> File.read!() |> String.trim(),
      # Informational
      name: @name,
      description: @description,
      source_url: @github_url,
      homepage_url: @homepage_url,
      # Configuration
      aliases: aliases(),
      deps: deps(),
      docs: docs(),
      dialyzer: dialyzer(),
      package: package(),
      preferred_cli_env: preferred_cli_env(),
      test_coverage: test_coverage()
    ]

  def application,
    do: [
      mod: {Matcha.Application, []},
      extra_applications: extra_applications(Mix.env())
    ]

  defp extra_applications(:prod), do: []

  defp extra_applications(_env),
    do: [
      :dialyzer,
      :mnesia
    ]

  defp aliases,
    do: [
      # Benchmark report generation
      benchmarks: [
        "test --include benchmark",
        "benchmarks.index"
      ],
      "benchmarks.index": &index_benchmarks/1,
      # Combination check utility
      checks: [
        "test.suites",
        "lint",
        "typecheck"
      ],
      # Combination clean utility
      clean: [
        "typecheck.clean",
        "deps.clean --all",
        &clean_build_folders/1
      ],
      # Coverage report generation
      coverage: "coveralls.html #{@default_test_suite_includes}",
      # Documentation tasks
      "docs.coverage": "doctor",
      # "docs.coverage": "inch",
      # "docs.coverage.report": "inch.report",
      # Mix installation tasks
      install: [
        "install.rebar",
        "install.hex",
        "install.deps"
      ],
      "install.rebar": "local.rebar --force",
      "install.hex": "local.hex --force",
      "install.deps": "deps.get",
      # Linting tasks
      lint: [
        "lint.compile",
        "lint.format",
        "lint.style"
      ],
      "lint.compile": "compile --force --warnings-as-errors",
      "lint.format": "format --check-formatted",
      "lint.style": "credo --strict",
      # Release tasks
      release: "hex.publish",
      # Static pages tasks
      static: [
        "benchmarks",
        "coverage",
        "docs",
        "static.collect"
      ],
      "static.collect": &collect_static_pages/1,
      # Typecheck tasks
      typecheck: [
        "typecheck.run"
      ],
      "typecheck.build-cache": "dialyzer --plt --format dialyxir",
      "typecheck.clean": "dialyzer.clean",
      "typecheck.explain": "dialyzer.explain --format dialyxir",
      "typecheck.run": "dialyzer --format dialyxir",
      # Test tasks
      "test.benchmarks": "test --include benchmark",
      "test.doctest": "test --include doctest",
      "test.usage": "test --include usage",
      "test.unit": "test --include unit",
      # run only default test suites
      "test.suites": "test #{@default_test_suite_includes}",
      # coverage for everything but benchmarks
      "test.coverage": "coveralls #{@default_test_suite_includes}",
      "test.coverage.report": "coveralls.github #{@default_test_suite_includes}"
    ]

  defp deps,
    do: [
      {:recon, ">= 2.3.0"},
      # Dev tooling
      {:benchee, "~> 1.0", only: @dev_envs, runtime: false},
      {:benchee_html, "~> 1.0", only: @dev_envs, runtime: false},
      {:credo, "~> 1.6", only: @dev_envs, runtime: false},
      {:dialyxir, "~> 1.0", only: @dev_envs, runtime: false},
      {:doctor, "~> 0.21", only: @dev_envs, runtime: false},
      {:ex_doc, "~> 0.29", only: @dev_envs, runtime: false},
      {:excoveralls, "~> 0.14 and >= 0.14.4", only: @dev_envs},
      {:jason, ">= 0.0.1", only: @dev_envs, runtime: false}
    ]

  defp docs,
    do: [
      # Metadata
      name: @name,
      authors: @authors,
      source_ref: @release_branch,
      source_url: @github_url,
      homepage_url: @homepage_url,
      # Files and Layout
      extra_section: "OVERVIEW",
      main: "Matcha",
      logo: "docs/img/logo.png",
      cover: "docs/img/cover.png",
      extras: [
        # Guides
        "docs/guides/usage.livemd": [filename: "guide-usage", title: "Using Matcha"],
        "docs/guides/usage/filtering-and-mapping.livemd": [
          filename: "guide-filtering-and-mapping",
          title: "...for Filtering/Mapping"
        ],
        "docs/guides/usage/tables.livemd": [
          filename: "guide-tables",
          title: "...for ETS/DETS/Mnesia"
        ],
        "docs/guides/usage/tracing.livemd": [
          filename: "guide-tracing",
          title: "...for Tracing"
        ],
        "docs/guides/adoption.livemd": [filename: "guide-adoption", title: "Adopting Matcha"],
        # Cheatsheets
        "docs/cheatsheets/adoption.cheatmd": [
          filename: "cheatsheet-adoption",
          title: "Adoption Cheatsheet"
        ],
        "docs/cheatsheets/tracing.cheatmd": [
          filename: "cheatsheet-tracing",
          title: "Tracing Cheatsheet"
        ],
        # Reference
        "CHANGELOG.md": [filename: "changelog", title: "Changelog"],
        "CONTRIBUTING.md": [filename: "contributing", title: "Contributing"],
        "CONTRIBUTORS.md": [filename: "contributors", title: "Contributors"],
        "LICENSE.md": [filename: "license", title: "License"]
      ],
      groups_for_extras: [
        Guides: ~r|docs/guides|,
        Cheatsheets: ~r|docs/cheatsheets|,
        Reference: [
          "CHANGELOG.md",
          "CONTRIBUTING.md",
          "CONTRIBUTORS.md",
          "LICENSE.md"
        ]
      ],
      groups_for_modules: [
        Core: [
          Matcha,
          Matcha.Pattern,
          Matcha.Spec
        ],
        Tables: [
          Matcha.Table,
          Matcha.Table.ETS,
          Matcha.Table.ETS.Match,
          Matcha.Table.ETS.Select,
          Matcha.Table.Mnesia,
          Matcha.Table.Mnesia.Match,
          Matcha.Table.Mnesia.Select
        ],
        Tracing: [
          Matcha.Trace,
          Matcha.Trace.Calls,
          Matcha.Trace.Messages,
          Matcha.Trace.Processes
        ],
        Exceptions: [
          Matcha.Error,
          Matcha.Pattern.Error,
          Matcha.Rewrite.Error,
          Matcha.Spec.Error,
          Matcha.Trace.Error
        ],
        Internals: [
          Matcha.Context,
          Matcha.Context.Erlang,
          Matcha.Context.FilterMap,
          Matcha.Context.Match,
          Matcha.Context.Table,
          Matcha.Context.Trace,
          Matcha.Rewrite,
          Matcha.Rewrite.Kernel,
          Matcha.Source
        ]
      ],
      nest_modules_by_prefix: [
        # Matcha.Context,
        # Matcha.Table,
        # Matcha.Trace,
        # Matcha.Error
      ]
    ]

  # Control dialyzer success-typing engine
  defp dialyzer,
    do: [
      plt_file: {:no_warn, "priv/plts/dialyzer.plt"},
      flags:
        ["-Wunmatched_returns", :error_handling, :underspecs] ++
          if :erlang.system_info(:otp_release) != '25' do
            [:race_conditions]
          else
            []
          end,
      ignore_warnings: ".dialyzer_ignore.exs",
      list_unused_filters: true,
      plt_add_apps: [:mnesia],
      plt_ignore_apps: []
    ]

  # Hex.pm information
  defp package,
    do: [
      maintainers: @maintainers,
      licenses: @licenses,
      links: %{
        Home: @homepage_url,
        GitHub: @github_url
      },
      files: [
        "lib",
        "mix.exs",
        "CHANGELOG.md",
        "CONTRIBUTING.md",
        "LICENSE.md",
        "README.md",
        "VERSION"
      ]
    ]

  defp preferred_cli_env,
    do: aliases() |> Keyword.keys() |> Enum.map(fn alias -> {alias, :test} end)

  defp test_coverage,
    do: [
      tool: ExCoveralls
    ]

  defp clean_build_folders(_) do
    ~w[_build bench cover deps doc] |> Enum.map(&File.rm_rf!/1)
  end

  defp index_benchmarks(_) do
    IO.puts("Creating bench/index.html...")

    list_items =
      Path.wildcard("bench/*/**/*.html")
      |> Enum.map(fn html_file ->
        relative_file = String.replace_leading(html_file, "bench/", "")
        ~s|<li><a href="#{relative_file}">Benchmark: #{relative_file}</a></li>|
      end)

    index_html = ["<ul>", list_items, "</ul>"]

    File.write!("bench/index.html", index_html)
  end

  defp collect_static_pages(_) do
    IO.puts("Collecting static files under static/...")

    File.mkdir_p!("static")

    File.cp_r!("bench", "static/bench")

    File.mkdir_p!("static/cover")
    File.cp!("cover/excoveralls.html", "static/cover/index.html")

    File.cp_r!("doc", "static/doc")

    File.write!("static/index.html", ~s|
      <ul>
        <li><a href="bench/index.html">Benchmarks</a></li>
        <li><a href="cover/index.html">Coverage</a></li>
        <li><a href="doc/index.html">Documentation</a></li>
    |)
  end
end
